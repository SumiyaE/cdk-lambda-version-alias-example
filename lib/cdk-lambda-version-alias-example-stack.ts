import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as sqs from 'aws-cdk-lib/aws-sqs';
import {NodejsFunction} from "aws-cdk-lib/aws-lambda-nodejs";
import {Alias, Runtime, Version} from "aws-cdk-lib/aws-lambda";
import {SqsEventSource} from "aws-cdk-lib/aws-lambda-event-sources";

export class CdkLambdaVersionAliasExampleStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // lambdaの作成
    const lambda = new NodejsFunction(this, 'CdkLambdaVersionAliasExampleFunction', {
      runtime: Runtime.NODEJS_LATEST,
      entry: 'lambda/handler.ts',
      handler: 'handler',
    });
    // lambdaの$Latestのバージョンを取得
    const lambdaLatestVersion = lambda.latestVersion
    // lambdaのバージョンを発行
    const lambdaUniqueVersion =  new Version(this, 'CdkLambdaVersionAliasExampleVersion', {
      lambda: lambda
    })

    // lambdaのエイリアスを作成
    const alias = new Alias(this, 'CdkLambdaVersionAliasExampleAlias', {
      aliasName: 'test',
      version: lambdaLatestVersion,
      additionalVersions: [{
        version: lambdaUniqueVersion,
        weight: 0.5
      }]
    });


    // SQSの作成
    const queue = new sqs.Queue(this, 'CdkLambdaVersionAliasExampleQueue', {
      visibilityTimeout: cdk.Duration.seconds(300)
    });

    // SQSをLambdaのトリガーに設定
    lambda.addEventSource(new SqsEventSource(queue));
  }
}
