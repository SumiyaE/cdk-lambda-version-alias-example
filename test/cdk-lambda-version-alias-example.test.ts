import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as CdkLambdaVersionAliasExample from '../lib/cdk-lambda-version-alias-example-stack';
import {expect, test} from "vitest";
import {ignoreAssetHashSerializer} from "./plugins/ignore-asset-hash";
// example test. To run these tests, uncomment this file along with the
// example resource in lib/cdk-lambda-version-alias-example-stack.ts
test('SQS Queue Created', () => {
  const app = new cdk.App();
  const stack = new CdkLambdaVersionAliasExample.CdkLambdaVersionAliasExampleStack(app, 'MyTestStack');
  const template = Template.fromStack(stack);

  // snapshot testを実行する
  // s3のbucket名やlambdaのコードのhash値が毎回変わるため、それらを無視するためのserializerを追加
  expect.addSnapshotSerializer(ignoreAssetHashSerializer);
  expect(template).toMatchSnapshot();
});
