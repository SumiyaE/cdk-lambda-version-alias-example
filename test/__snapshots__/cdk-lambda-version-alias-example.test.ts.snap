// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SQS Queue Created 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CdkLambdaVersionAliasExampleAlias6E7DC8B0": {
      "Properties": {
        "FunctionName": {
          "Ref": "CdkLambdaVersionAliasExampleFunction8E5C8A1A",
        },
        "FunctionVersion": "$LATEST",
        "Name": "test",
        "RoutingConfig": {
          "AdditionalVersionWeights": [
            {
              "FunctionVersion": {
                "Fn::GetAtt": [
                  "CdkLambdaVersionAliasExampleVersion0CB4103F",
                  "Version",
                ],
              },
              "FunctionWeight": 0.5,
            },
          ],
        },
      },
      "Type": "AWS::Lambda::Alias",
    },
    "CdkLambdaVersionAliasExampleFunction8E5C8A1A": {
      "DependsOn": [
        "CdkLambdaVersionAliasExampleFunctionServiceRoleDefaultPolicyBD439B03",
        "CdkLambdaVersionAliasExampleFunctionServiceRole87861BA8",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "CdkLambdaVersionAliasExampleFunctionServiceRole87861BA8",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "CdkLambdaVersionAliasExampleFunctionServiceRole87861BA8": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CdkLambdaVersionAliasExampleFunctionServiceRoleDefaultPolicyBD439B03": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "CdkLambdaVersionAliasExampleQueue3D8ED1F9",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CdkLambdaVersionAliasExampleFunctionServiceRoleDefaultPolicyBD439B03",
        "Roles": [
          {
            "Ref": "CdkLambdaVersionAliasExampleFunctionServiceRole87861BA8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CdkLambdaVersionAliasExampleFunctionSqsEventSourceMyTestStackCdkLambdaVersionAliasExampleQueue095E73199C51880A": {
      "Properties": {
        "EventSourceArn": {
          "Fn::GetAtt": [
            "CdkLambdaVersionAliasExampleQueue3D8ED1F9",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "CdkLambdaVersionAliasExampleFunction8E5C8A1A",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "CdkLambdaVersionAliasExampleQueue3D8ED1F9": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "CdkLambdaVersionAliasExampleVersion0CB4103F": {
      "Properties": {
        "FunctionName": {
          "Ref": "CdkLambdaVersionAliasExampleFunction8E5C8A1A",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
